test/UnitTests/GitHub.VisualStudio/Services/MetricsTests.cs

```diff
@@ -0,0 +1,506 @@
+                NumberOfClones = 1,
+            }, sameDay: false, sameWeek: false, sameMonth: false);
+            Func<Task> tick = null;
+
+            service.WhenForAnyArgs(x => x.StartTimer(null, new TimeSpan(), new TimeSpan()))
+                .Do(x => tick = x.ArgAt<Func<Task>>(0));
+
+            var serviceProvider = CreateServiceProvider();
+            var target = new UsageTracker(serviceProvider, service);
+
+            await tick();
+
+            await service.Received().WriteLocalData(
+                Arg.Is<UsageData>(x =>
+                    x.Model.NumberOfStartups == 0 &&
+                    x.Model.NumberOfStartupsWeek == 0 &&
+                    x.Model.NumberOfStartupsMonth == 0 &&
+                    x.Model.NumberOfClones == 0));
+        }
+
+        [Test]
+        public async Task ShouldIncrementCounter()
+        {
+            var model = new UsageModel { NumberOfClones = 4 };
+            var usageService = CreateUsageService(model);
+            var target = new UsageTracker(
+                CreateServiceProvider(),
+                usageService);
+
+            await target.IncrementCounter(x => x.NumberOfClones);
+            UsageData result = usageService.ReceivedCalls().First(x => x.GetMethodInfo().Name == "WriteLocalData").GetArguments()[0] as UsageData;
+
+            Assert.AreEqual(5, result.Model.NumberOfClones);
+        }
+
+        [Test]
+        public async Task ShouldWriteUpdatedData()
+        {
+            var data = new UsageData { Model = new UsageModel() };
+            var service = CreateUsageService(data);
+            var target = new UsageTracker(
+                CreateServiceProvider(),
+                service);
+
+            await target.IncrementCounter(x => x.NumberOfClones);
+
+            await service.Received(1).WriteLocalData(data);
+        }
+
+        [Test]
+        public async Task UsageServiceWritesAllTheDataCorrectly()
+        {
+            var model = CreateUsageModel();
+            var serviceProvider = CreateServiceProvider();
+            var usageService = CreateUsageService(model, sameDay: true);
+            var targetAndTick = CreateTargetAndGetTick(serviceProvider, usageService);
+            var vsservices = serviceProvider.GetService<IVSServices>();
+            vsservices.VSVersion.Returns(model.VSVersion);
+
+            await targetAndTick.Item2();
+
+            var metricsService = serviceProvider.TryGetService<IMetricsService>();
+
+            var expected = model;
+            expected.NumberOfStartups++;
+            expected.NumberOfStartupsWeek++;
+            expected.NumberOfStartupsMonth++;
+
+            var result = (usageService.ReceivedCalls().First(x => x.GetMethodInfo().Name == "WriteLocalData").GetArguments()[0] as UsageData).Model;
+            CollectionAssert.AreEquivalent(
+                ReflectionUtils.GetProperties(expected.GetType()).Select(x => new { x.Name, Value = x.GetValue(expected) }),
```

Comment:
One last thought about this. While reading https://msdn.microsoft.com/en-us/library/ms182276.aspx I noticed:

> For value types, the inherited implementation of Equals uses the Reflection library, and compares the contents of all fields.

So we could actually just do `Assert.That(actual, Is.EqualTo(expected))` here.

Reply:
Equals won't say which fields are mismatched, which makes it really hard to figure out what's wrong from the test.